File: sdk/c/src/providers/tempmail_la.c
/**
 * tempmail.la 渠道实现（支持分页）
 */
#include "tempmail_internal.h"

#define TLA_BASE "https://tempmail.la/api"

static const char* tla_headers[] = {
    "Content-Type: application/json",
    "accept-language: zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "cache-control: no-cache",
    "dnt: 1",
    "locale: zh-CN",
    "origin: https://tempmail.la",
    "platform: PC",
    "pragma: no-cache",
    "product: TEMP_MAIL",
    "referer: https://tempmail.la/zh-CN/tempmail",
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0",
    "sec-ch-ua: \"Not(A:Brand\";v=\"8\", \"Chromium\";v=\"144\", \"Microsoft Edge\";v=\"144\"",
    "sec-ch-ua-mobile: ?0",
    "sec-ch-ua-platform: \"Windows\"",
    "sec-fetch-dest: empty",
    "sec-fetch-mode: cors",
    "sec-fetch-site: same-origin",
    NULL
};

tm_email_info_t* tm_provider_tempmail_la_generate(void) {
    tm_http_response_t *resp = tm_http_request(TM_HTTP_POST, TLA_BASE "/mail/create", tla_headers, "{\"turnstile\":\"\"}", 15);
    if (!resp || resp->status != 200) { tm_http_response_free(resp); return NULL; }

    cJSON *json = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!json) return NULL;

    cJSON *code = cJSON_GetObjectItemCaseSensitive(json, "code");
    if (!code || code->valueint != 0) { cJSON_Delete(json); return NULL; }

    cJSON *data = cJSON_GetObjectItemCaseSensitive(json, "data");
    tm_email_info_t *info = tm_email_info_new();
    info->channel = CHANNEL_TEMPMAIL_LA;
    info->email = tm_strdup(cJSON_GetStringValue(cJSON_GetObjectItemCaseSensitive(data, "address")));
    cJSON *exp = cJSON_GetObjectItemCaseSensitive(data, "endAt");
    info->expires_at = exp && cJSON_IsNumber(exp) ? (long long)exp->valuedouble : 0;
    cJSON_Delete(json);
    return info;
}

tm_email_t* tm_provider_tempmail_la_get_emails(const char *email, int *count) {
    *count = -1;

    /* 收集所有分页邮件 */
    cJSON *all = cJSON_CreateArray();
    char *cursor = NULL;

    while (1) {
        char body[512];
        if (cursor) snprintf(body, sizeof(body), "{\"address\":\"%s\",\"cursor\":\"%s\"}", email, cursor);
        else snprintf(body, sizeof(body), "{\"address\":\"%s\",\"cursor\":null}", email);

        tm_http_response_t *resp = tm_http_request(TM_HTTP_POST, TLA_BASE "/mail/box", tla_headers, body, 15);
        if (!resp || resp->status != 200) { tm_http_response_free(resp); break; }

        cJSON *json = cJSON_Parse(resp->body);
        tm_http_response_free(resp);
        if (!json) break;

        cJSON *code = cJSON_GetObjectItemCaseSensitive(json, "code");
        if (!code || code->valueint != 0) { cJSON_Delete(json); break; }

        cJSON *data = cJSON_GetObjectItemCaseSensitive(json, "data");
        cJSON *rows = cJSON_GetObjectItemCaseSensitive(data, "rows");
        if (cJSON_IsArray(rows)) {
            int n = cJSON_GetArraySize(rows);
            for (int i = 0; i < n; i++) {
                cJSON_AddItemToArray(all, cJSON_Duplicate(cJSON_GetArrayItem(rows, i), 1));
            }
        }

        int has_more = cJSON_IsTrue(cJSON_GetObjectItemCaseSensitive(data, "hasMore"));
        const char *next = cJSON_GetStringValue(cJSON_GetObjectItemCaseSensitive(data, "cursor"));
        free(cursor);
        cursor = NULL;
        if (has_more && next) cursor = tm_strdup(next);
        cJSON_Delete(json);
        if (!has_more || !cursor) break;
    }

    free(cursor);
    int n = cJSON_GetArraySize(all);
    *count = n;
    if (n == 0) { cJSON_Delete(all); return NULL; }

    tm_email_t *emails = tm_emails_new(n);
    for (int i = 0; i < n; i++) emails[i] = tm_normalize_email(cJSON_GetArrayItem(all, i), email);
    cJSON_Delete(all);
    return emails;
}

File: sdk/rust/src/providers/tempmail_la.rs
/*!
 * tempmail.la 渠道实现（支持分页）
 */

use reqwest::blocking::Client;
use serde_json::Value;
use crate::types::{Channel, EmailInfo, Email};
use crate::normalize::normalize_email;

const BASE_URL: &str = "https://tempmail.la/api";

fn client() -> Client {
    Client::builder()
        .user_agent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Safari/537.36 Edg/144.0.0.0")
        .timeout(std::time::Duration::from_secs(15))
        .build().unwrap()
}

pub fn generate_email() -> Result<EmailInfo, String> {
    let resp = client()
        .post(format!("{}/mail/create", BASE_URL))
        .header("Content-Type", "application/json")
        .header("accept-language", "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6")
        .header("cache-control", "no-cache")
        .header("dnt", "1")
        .header("locale", "zh-CN")
        .header("origin", "https://tempmail.la")
        .header("platform", "PC")
        .header("pragma", "no-cache")
        .header("product", "TEMP_MAIL")
        .header("referer", "https://tempmail.la/zh-CN/tempmail")
        .header("sec-ch-ua", "\"Not(A:Brand\";v=\"8\", \"Chromium\";v=\"144\", \"Microsoft Edge\";v=\"144\"")
        .header("sec-ch-ua-mobile", "?0")
        .header("sec-ch-ua-platform", "\"Windows\"")
        .header("sec-fetch-dest", "empty")
        .header("sec-fetch-mode", "cors")
        .header("sec-fetch-site", "same-origin")
        .json(&serde_json::json!({"turnstile": ""}))
        .send().map_err(|e| format!("tempmail-la request failed: {}", e))?;

    if !resp.status().is_success() {
        return Err(format!("tempmail-la generate failed: {}", resp.status()));
    }

    let data: Value = resp.json().map_err(|e| format!("parse failed: {}", e))?;
    if data["code"].as_i64() != Some(0) {
        return Err("Failed to generate email".into());
    }

    Ok(EmailInfo {
        channel: Channel::TempmailLa,
        email: data["data"]["address"].as_str().unwrap_or("").to_string(),
        token: None,
        expires_at: data["data"]["endAt"].as_i64(),
        created_at: data["data"]["startAt"].as_str().map(|s| s.to_string()),
    })
}

pub fn get_emails(email: &str) -> Result<Vec<Email>, String> {
    let mut all_emails = Vec::new();
    let mut cursor: Option<String> = None;

    loop {
        let resp = client()
            .post(format!("{}/mail/box", BASE_URL))
            .header("Content-Type", "application/json")
            .header("dnt", "1")
            .header("origin", "https://tempmail.la")
            .header("platform", "PC")
            .header("product", "TEMP_MAIL")
            .header("referer", "https://tempmail.la/zh-CN/tempmail")
            .header("sec-ch-ua", "\"Not(A:Brand\";v=\"8\", \"Chromium\";v=\"144\", \"Microsoft Edge\";v=\"144\"")
            .header("sec-ch-ua-mobile", "?0")
            .header("sec-ch-ua-platform", "\"Windows\"")
            .header("sec-fetch-dest", "empty")
            .header("sec-fetch-mode", "cors")
            .header("sec-fetch-site", "same-origin")
            .json(&serde_json::json!({"address": email, "cursor": cursor}))
            .send().map_err(|e| format!("tempmail-la request failed: {}", e))?;

        if !resp.status().is_success() {
            return Err(format!("tempmail-la get emails failed: {}", resp.status()));
        }

        let data: Value = resp.json().map_err(|e| format!("parse failed: {}", e))?;
        if data["code"].as_i64() != Some(0) {
            return Err("Failed to get emails".into());
        }

        if let Some(rows) = data["data"]["rows"].as_array() {
            for raw in rows {
                all_emails.push(normalize_email(raw, email));
            }
        }

        if data["data"]["hasMore"].as_bool() == Some(true) {
            if let Some(c) = data["data"]["cursor"].as_str() {
                cursor = Some(c.to_string());
                continue;
            }
        }
        break;
    }

    Ok(all_emails)
}
